appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardContextValve.java

```diff
@@ -233,6 +234,75 @@ private void log(String message, Throwable t) {
     }
     */
 
+    /**
+     * resolves '.' and '..' elements in the path
+     * if there are too many, making a path negative, returns null
+     *
+     * @param path to be normalized
+     * @return normalized path or null
+     */
+    protected String normalize(String path) {
+        if (path == null) {
+            return null;
+        }
+
+        String rv = path;
+        // starts with a double-slash
+        if(rv.indexOf("//") == 0) {
+            rv = rv.replace("//", "/");
+        }
+        // starts with dot-slash
+        if(rv.indexOf("./") == 0) {
+            rv = rv.replaceFirst("./", "/");
+        }
+
+        // Normalize the slashes and add leading slash if necessary
+        if (rv.indexOf('\\') >= 0) {
+            rv = rv.replace('\\', '/');
+        }
+
+        // Resolve occurrences of "/../" in the normalized path
+        while (true) {
+            int idx = rv.indexOf("/../");
+            if (idx < 0) {
+                break;
+            }
+            if (idx == 0) {
+                return null;  // negative relative path
+            }
+            int index2 = rv.lastIndexOf('/', idx - 1);
+            rv = rv.substring(0, index2) + rv.substring(idx + 3);
+        }
+
+        // Resolve occurrences of "/./" example if the path looks like /app/./some/./something/./my.jsp
+        //then after processing should look like /app/some/something/my.jsp
+        rv = evaluateNormalizedPathWithSinglePoint(rv);
+
+        //if the path don't start with / then include it
+        if(!rv.startsWith("/")) {
+            rv = "/" + rv;
+        }
+
+        // Return the normalized path that we have completed
+        return rv;
+    }
+
+    /**
+     * this method helps to evaluate the element "/./" on the path
+     * @param path to be normalized
+     * @return normalized path
+     */
+    private String evaluateNormalizedPathWithSinglePoint(String path) {
```

Comment:
I think this could be `static`, package level and unit tested.

Reply:
Done!

Reply:
Doing it in #24215.

Reply:
Doing it in #24216 (UT).